jax/experimental/export/_shape_poly_decision.py

```diff
@@ -0,0 +1,412 @@
+          e_minus_m_ub = _DimExpr._merge_sorted_terms(e_minus_m, 0, 1,
+                                                      [(_DimMon(), 1)], 0, e_c * int(m_ub))
+          acc_combination(_DimExpr(dict(e_minus_m_ub), e.scope))
+      else:
+        if m_lb > -np.inf:
+          e_minus_m = _DimExpr._merge_sorted_terms(e._monomials_sorted, 0, 1,
+                                                   [(e_m, e_c)], 0, -1)
+          e_minus_m_lb = _DimExpr._merge_sorted_terms(e_minus_m, 0, 1,
+                                                      [(_DimMon(), 1)], 0, e_c * int(m_lb))
+          acc_combination(_DimExpr(dict(e_minus_m_lb), e.scope))
+
+    for prev_constraints in self._expr_constraints.values():
+      for _, prev in prev_constraints:
+        # Compose "e" with "prev" if they have one monomial with different
+        # signs
+        for e_m, e_c in e.monomials():
+          if e_m.degree == 0: continue
+          prev_c = prev._coeffs.get(e_m)
+          if prev_c is not None and prev_c * e_c < 0:
+            new_constraint = _DimExpr(
+                dict(_DimExpr._merge_sorted_terms(e._monomials_sorted, 0, abs(prev_c),
+                                                  prev._monomials_sorted, 0, abs(e_c))),
+                e.scope)
+            acc_combination(new_constraint)
+            break
+
+    return combinations
+
+  def _add_to_state(self, e: _DimExpr,
+                    constraint_str: str):
+    """Updates the internal state to reflect "e >= 0". """
+    assert _DimExpr.to_constant(e) is None
+    for m, m_c in e.monomials():
+      if m.degree == 0: continue
+      _add_internal_constraints(self, m, e.scope)
+
+    if (mon_factors := e.to_single_term()) is not None:
+      n, mon_c, mon = mon_factors
+      bounds = self._term_bounds.get(mon, (- np.inf, np.inf))
+      if mon_c > 0:
+        mon_ge = int(np.ceil(- n / mon_c))
+        new_bounds = (max(mon_ge, bounds[0]), bounds[1])
+      else:
+        le = int(np.floor(-n / mon_c))
+        new_bounds = (bounds[0], min(le, bounds[1]))
+      if new_bounds[0] > new_bounds[1]:
+        raise ValueError(f"Unsatisfiable constraint: {constraint_str}")
+
+      self._term_bounds[mon] = new_bounds
+      return
+
+    lead_m, lead_m_c = e.leading_term
+    self._expr_constraints[lead_m].add((lead_m_c, e))
+
+  def bounds(self, e: DimSize,
+             stop_early: Callable[[float, float], bool] | None
+             ) -> tuple[float, float]:
+    """Returns the lower and upper bounds, or -+inf.
+
+    See more details in `_shape_poly.bounds_decision`.
+    """
+    if (const := _DimExpr.to_constant(e)) is not None:
+      return (const, const)
+    assert isinstance(e, _DimExpr)
+    cache_key = (e, stop_early)
+    if (res := self.scope._bounds_cache.get(cache_key)) is not None: return res
+    res = self._bounds_for_sorted_terms(e.scope, e._monomials_sorted, 0, stop_early)
+    lb, ub = res
+    lb = lb if np.isinf(lb) else int(lb)
```

Comment:
Why convert the bound to an `int` here?

Reply:
Actually we do not need the conversion to `int`. (The bounds are declared as `float` but they are really `int | np.inf | -np.inf`, but that is not important here.